spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/ThreadChannelConnectionFactory.java

```diff
@@ -144,6 +150,40 @@ public synchronized void destroy() {
 		}
 	}
 
+	/**
+	 * Call to prepare to switch the channel(s) owned by this thread to another thread.
+	 * @return an opaque object representing the context to switch.
+	 * @since 2.3.7
+	 * @see #switchContext(Object)
+	 */
+	public Object prepareSwitchContext() {
+		return prepareSwitchContext(UUID.randomUUID());
+	}
+
+	Object prepareSwitchContext(UUID uuid) {
+		if (getPublisherConnectionFactory() instanceof ThreadChannelConnectionFactory) {
+			((ThreadChannelConnectionFactory) getPublisherConnectionFactory()).prepareSwitchContext(uuid);
+		}
+		this.contextSwitches.put(uuid, ((ConnectionWrapper) createConnection()).prepareSwitchContext());
+		return uuid;
```

Comment:
This all looks cool: very interesting and smart approach.

I can merge it as when you answer some my questions.

How about just to return a `Context` as is without storing in between into that `contextSwitches` map?
I wonder if some one will call `prepareSwitchContext()`, but won't call `switchContext()`... So, feels like a memory leak.

At least we should consider to empty that map on CF destroy.

Also it looks like when we call `prepareSwitchContext()` several times, we end up with empty `Context`, just because we called `this.channels.remove()` for the first one. Probably better to protect ourselves just only for a one call.

And then `switchContext(Context context)` against an empty one will finish just with void.

Does it ring a bell?

Thanks

Reply:
I thought about just returning a `Context` but the problem is that we have a nested `publisherConnectionFactory` which could be (and is by default) another TCCF, and the `RabbitTemplate` uses it by default.

That means we need a `Context` for each factory bound to the same key.

I also didn't want to make it easy for them to reflect into the `Context` and access the channels directly.

I will push another commit with some of your other suggestions to add some more protection, but if a user does something like never call  `switchContext` there's not really much we can do about that.